presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java

```diff
@@ -0,0 +1,282 @@
+            }
+            return assignmentsBuilder.build();
+        }
+
+        private List<InPredicate> collectInPredicates(List<Expression> expressions)
+        {
+            ImmutableList.Builder<InPredicate> listBuilder = ImmutableList.builder();
+            for (Expression expression : expressions) {
+                new DefaultTraversalVisitor<Void, Void>()
+                {
+                    @Override
+                    protected Void visitInPredicate(InPredicate node, Void context)
+                    {
+                        listBuilder.add(node);
+                        return null;
+                    }
+                }.process(expression, null);
+            }
+            return listBuilder.build();
+        }
+    }
+
+    private class InsertSemiJoinRewriter
+            extends SimplePlanRewriter<InsertSemiJoinContext>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final SymbolAllocator symbolAllocator;
+
+        public InsertSemiJoinRewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode node, RewriteContext<InsertSemiJoinContext> context)
+        {
+            InPredicate inPredicate = context.get().peekInPredicate();
+            InPredicate unprojectedInPredicate = unprojectInPredicate(node, inPredicate);
+            if (!inPredicate.equals(unprojectedInPredicate)) {
+                context.get().pushInPredicate(unprojectedInPredicate);
+            }
+            node = (ProjectNode) super.visitProject(node, context);
+            Optional<Symbol> inPredicateSymbol = context.get().getInPredicateSymbol();
+            if (inPredicateSymbol.isPresent()) {
+                return tryAppendProjection(node, inPredicateSymbol.get());
+            }
+            else {
+                return node;
+            }
+        }
+
+        private InPredicate unprojectInPredicate(ProjectNode node, InPredicate inPredicate)
+        {
+            Expression value = inPredicate.getValue();
+            Expression valueList = inPredicate.getValueList();
+
+            for (Symbol assignmentKey : node.getAssignments().keySet()) {
+                Expression assignmentValue = node.getAssignments().get(assignmentKey);
+                Expression assigmentKeyExpression = assignmentKey.toQualifiedNameReference();
+                if (value.equals(assigmentKeyExpression)) {
+                    value = assignmentValue;
+                }
+                if (valueList.equals(assigmentKeyExpression)) {
+                    valueList = assignmentValue;
+                }
+                inPredicate = new InPredicate(value, valueList);
```

Comment:
do not construct `InPredicate` here.
Instead just replace `return inPredicate` after the loop with `return new InPredicate(value, valueList)`.